man/ippon.8
.TH ippon 8 "Wed Jul 14 2004" "" "Network UPS Tools (NUT)"
.SH NAME
ippon \- Driver for Ippon UPS equipment
.SH NOTE
This driver is obsolete and has been replaced by the \fBmegatec\fR(8)
driver. It will be removed somewhere in the near future.

This man page only documents the hardware\(hyspecific features of the
ippon driver.  For information about the core driver, see
\fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
This driver supports Ippon UPS equipment.

.SH EXTRA ARGUMENTS

This driver supports the following optional settings in the
\fBups.conf\fR(5):

.IP "sdelay=num"

Specify delay of num minutes before switching off the load.

.IP "rdelay=num"

Specify delay of num minutes before restoring power (0 = never).

.SH AUTHOR
Alexander Fedorov <alexf@land.ru>

.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.5 - what's new since 2.0.4:

 This release is a backport of the development version.  Many changes
 have already been backported previously.  Thus it is more a
 synchronisation release, though it includes many bugfixes and support
 for new models.

 - support for new devices: APC Smart-UPS with 6TI firmware; Belkin
   Small Enterprise F6C1500-TW-RK; Compaq R3000 XR, R5500 XR; Cyber
   Power 550SL, 725SL, 685AVR, 800AVR, 1200AVR, AE550; Eltek; Inform
   GUARD; Microsol Rhino; Opti-UPS PowerES 420E; PowerMan RealSmart,
   BackPro; Powerware PW9315 3-phase; SOLA 305; Tripp-Lite
   SMART550USB, SMART2200RMXL2U, OMNI1000LCD, OMNI900LCD, OMNI650LCD,
   1500 LCD, AVR550U; Viewsonic PowerES 420E.
 - bcmxcp: added 3-phase support
 - megatec: better hardware support, more instant commands
 - mge-hid: support more instant commands
 - newhidups: fixed APC and Tripp Lite bugs, various memory bugs,
   improved report buffering, improved Solaris support, added
   '-x explore' option for easy diagnosis of new devices
 - solis: shutdown programming, support new cables, Solaris support
 - tripplite_usb: updated SMARTPRO support, fixed OL/OB reporting, 
   better error handling, some memory bugs
 - new dummy-ups driver simulator
 - added HTML interface for access to CGI scripts

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.4 - what's new since 2.0.3:

 - The newhidups critical bug (segmentation fault) has been fixed. It has
   also received some more care, like buxfixes and new models support and
   enhancement for Solaris.
   [Peter Selinger and Arnaud Quette]

 - A bug has been fixed in NUT core to support resuming from suspend-to-disk.
   This should also fix other similar issues, like time synchronisation
   through the NTP - Network Time Protocol.
   [Arjen de Korte]

 - The mge-shut driver now better detects the Low Battery status, support
   new models and fixes some wrong status and data. It also fixes some
   issue where the UPS wasn't restarting (refer to mge-shut manpage).
   [Arnaud Quette]

 - The genericups custom configuration through ups.conf is working again
   [Arjen de Korte]

 - The genericups driver type 22 also support CyberPower 725SL
   (and maybe others SL models)
   [David Kaufman]

 - The new megatec driver, which will replace a bunch of drivers by nut 2.2
   (refer to docs/megatec.txt and UPGRADING) has been backported from the
   trunk (Development tree). The powermust driver has also received some
   attention.
   [Carlos Rodrigues]

 - The new rhino driver was added to support Microsol Rhino UPS hardware
   The solis has also been improved for solaris compatibility, and
   internal / external shutdown programming. solis can now save external
   shutdown programming to ups, and support new cables for solis 3
   [Silvino B. MagalhÃ£es]

 - Several fixes and improvements have been made to upsrw, upsset,
   cpsups, tripplite_usb and the FAQ.
   [Arjen de Korte and Charles Lepple]

---------------------------------------------------------------------------
Release notes for NUT 2.0.3 - what's new since 2.0.2:

 - The recent and major newhidups changes have been backported from the
   Development tree. It now:
   - supports models from MGE UPS SYSTEMS, APC and Belkin. Mustek and Unitek
     units are also recognized for development purpose,
   - handles better device reopening, after a disconnection,
   - handles multiple devices, with several parameters to find the right UPS.
   [Peter Selinger, Charles Lepple and Arnaud Quette]

 - The bcmxcp_usb driver has been added to support Powerware USB units.
   [Wolfgang Ocker and Kjell Claesson]

 - The tripplite_usb driver has been added to support Tripp Lite USB units.
   [Charles Lepple]

 - The sec driver is back as gamatronic
   [Gamatronic, Nadav Moskovitch]

 - The genericups driver has received official care from Gamatronic
   to add support for the Gamatronic UPS with alarm interface.
   [Gamatronic, Nadav Moskovitch]

 - The powermust driver now supports Soyntec Sekury C 500 and C 800 units.
   [Hanno Borns]

 - The mge-shut driver has received a bit of attention too, and enhance
   ups.model retrieval for some specific case (release 0.65)

 - The drivers don't change to the "statepath" directory anymore at
   initialisation time if called using -k. This avoid unneeded
   failure to poweroff the UPS if /var is already unmounted.
   [Gaspar Bakos]

 - The belkinunv driver now supports Belkin F6C1100-UNV
   [Dave Breiland]

 - The isbmex driver has been upgraded to version 0.05, which fixes
   various errors in formulas, add shutdown capability and revert
   back baudrate to B9600 (instead of B2400), as it broke the
   communication
   [Ricardo Martinezgarza]

 - The support of Sysgration UPGUARDS Pro650 in fentonups has
   been fixed
   [Simon J. Rowe]

 - The packaging files for Red Hat have received various fixes
   [Thomas Jarosch]

 - The solis driver has been fixed to avoid a naming colision and
   compile on Solaris
   [Paweł Kierdelewicz]

 - The snmp-ups driver has corrected the problem when exposing
   certain time data.

---------------------------------------------------------------------------
Release notes for NUT 2.0.2 - what's new since 2.0.1:

 - the newhidups USB driver has been improved a lot and is no more
   experimental. It also now has a basic APC support, which will
   soon replace the legacy hidups driver.

 - The mge-utalk driver has improved its support for old units.

 - The mge-shut driver has been improved for restart/shutdown
   sequences which was previously blocking the serial port.

 - The general MGE support has been added Pulsar EXtreme C / EX RT,

drivers/ippon.c
/* ippon.c - driver for Ippon UPS devices

   Copyright (C) 2004  Alexander Fedorov <alexf@land.ru>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#include "main.h"
#include "serial.h"
#include "ippon.h"

#define ENDCHAR		13	/* end == CR */
#define SECS            3       /* 3sec delay when get           */
#define USEC            0 	/* 0ms delay when get            */
#define SENDDELAY 	100000  /* 100ms between chars when send */
#define UPSDELAY 	150000  /* 150ms between send and get    */

static float lowvlt = 0,voltrange = 0,highvolt = 0;


static int instcmd(const char *cmdname, const char *extra)
{
	/*Start battery test*/
	if (!strcasecmp(cmdname,"test.battery.start"))
	{
	    ser_send_pace(upsfd,SENDDELAY,"T\r");
	    return STAT_INSTCMD_HANDLED;
	}
	
	/*Beeper on*/
	if (!strcasecmp(cmdname,"beeper.on"))
	{
	    ser_send_pace(upsfd,SENDDELAY,"Q\r");
	    return STAT_INSTCMD_HANDLED;
	}
	
	/*Beeper off*/
	if (!strcasecmp(cmdname,"beeper.off"))
	{
	    ser_send_pace(upsfd,SENDDELAY,"Q\r");
	    return STAT_INSTCMD_HANDLED;
	}
	
	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
	return STAT_INSTCMD_UNKNOWN;
}


void upsdrv_initinfo(void)
{
	dstate_setinfo("driver.version.internal", "%s", DRV_VERSION);
	dstate_setinfo("ups.mfr" , "Ippon");
	dstate_setinfo("ups.model" , "universal driver");
	
	dstate_addcmd("test.battery.start");
	dstate_addcmd("beeper.on");
	dstate_addcmd("beeper.off");
	
	upsh.instcmd = instcmd;
}

void upsdrv_updateinfo(void)
{
	char	buff[256],inpvlt[8],inpflt[8],outvlt[8],outcur[8],inpfrq[8],
		batvlt[8],temp[8],type[8];

	int 	ret;
	float 	bcharge;

	ret = ser_send_pace(upsfd,SENDDELAY,"Q1\r");
	if (ret != 3) 
	{
	    ser_comm_fail("Can't send status command");
	    dstate_datastale();
	    return;
	}

	usleep(UPSDELAY);
	
	ret = ser_get_line(upsfd,buff,sizeof(buff),ENDCHAR,"",SECS,USEC);
	
	if ( ret < 1 ) 
	{
	    ser_comm_fail("Can't read status");
	    dstate_datastale();
	    return;
	}
	
	if (strlen(buff) < 46)
	{
	    ser_comm_fail("Short read from UPS");
	    dstate_datastale();
	    return;
	}
	
	if (strlen(buff) > 46)
	{
	    ser_comm_fail("Long  read from UPS");
	    dstate_datastale();
	    return;
	}
	
	if (buff[0] !=  '(')
	{
	    ser_comm_fail("Wrong start char in sequence");
	    dstate_datastale();
	    return;
	}
	
	sscanf(buff,"%*c%s %s %s %s %s %s %s %s",inpvlt,inpflt,outvlt,outcur,
		inpfrq,batvlt,temp,type);
	
	bcharge = ((atof (batvlt) - 10.2) / 3.4) *100.0;
	
	if (bcharge > 100.0 ) bcharge = 100.0;

	dstate_setinfo("input.voltage" , "%s", inpvlt);
/*	dstate_setinfo("input.voltage.minimum", "%s", inpflt); */
	dstate_setinfo("output.voltage", "%s", outvlt);
	dstate_setinfo("ups.load", "%s" , outcur);
	dstate_setinfo("input.frequency", "%s" , inpfrq);
	dstate_setinfo("battery.charge", "%02.1f" , bcharge);
	dstate_setinfo("battery.voltage", "%s" , batvlt);
	dstate_setinfo("ups.temperature", "%s" , temp);
	
	status_init();
	
	if(type[0] == '0') 
	{
	    status_set("OL");
	}
	else status_set("OB");
	
	if (type[1] == '1') status_set("LB");    
	
	status_commit();
	dstate_dataok();
}

void upsdrv_shutdown(void)
{
	int sdelay = 1, rdelay = 1;
	char str[10];
	if (testvar("sdelay"))
		sdelay = atoi(getval("sdelay"));
	if (testvar("rdelay"))
		rdelay = atoi(getval("rdelay"));
	sprintf(str, "S%.2dR%.4d\r", sdelay, rdelay);
	ser_send_pace(upsfd, SENDDELAY, str); 
}




void upsdrv_help(void)
{
}

void upsdrv_makevartable(void)
{
	addvar(VAR_VALUE, "sdelay", "Specify delay before switching off the load");
	addvar(VAR_VALUE, "rdelay", "Specify delay before restoring power (0 - never)");
}

void upsdrv_banner(void)
{
	printf("Network UPS Tools - Ippon UPS driver %s (%s)\n\n", 
		DRV_VERSION, UPS_VERSION);

	upslogx(LOG_WARNING, "This driver is obsolete and has been replaced by the \"megatec\""
	                     " driver. It will be removed somewhere in the near future.");
}

void upsdrv_initups(void)
{
	char 	buff[256];
	int  	ret;
	int 	ratecurrent;
	float 	ratevolt,ratefreq;
	
	
	upsfd =	ser_open(device_path);
	ser_set_speed(upsfd,device_path,B2400);
/* we need send "F\r" twice.. and get answer only once,
   after second "F\r" sent (???)     				 */
	ret = ser_send_pace(upsfd,SENDDELAY,"F\r");
	usleep(UPSDELAY);
	ret = ser_get_line(upsfd,buff,sizeof(buff),ENDCHAR,"",SECS,USEC);
	
	ret = ser_send_pace(upsfd,SENDDELAY,"F\r");
	if (ret != 2) 
	{
	    ser_comm_fail("Can't send init command");
	    dstate_datastale();
	    return;
	}

	usleep(UPSDELAY);

	ret = ser_get_line(upsfd,buff,sizeof(buff),ENDCHAR,"",SECS,USEC);
	
	if ( ret < 1 ) 
	{
	    ser_comm_fail("Can't read init info");
	    dstate_datastale();
	    return;
	}

		
	if ((buff[0]=='#') && (strlen(buff) == 21)) 
	    {
		 printf("UPS identified\n") ;
	    }
	else 
	    {
		ser_comm_fail("Can't init UPS");
		dstate_datastale();
		return;
	    }
	
	sscanf(buff,"%*c %f %d %f %f",&ratevolt,&ratecurrent,&lowvlt,&ratefreq);
	upsdebugx(2, "UPS rated at %.2fV %dA %.2fHz \n ",ratevolt,ratecurrent,ratefreq);
	
	highvolt = 13.6;
	voltrange = highvolt - lowvlt;
	
}

void upsdrv_cleanup(void)
{
	ser_close(upsfd, device_path);
}
